File: data-import/getBeer.php
#!/usr/bin/php
<?php
ini_set("memory_limit","256M");

include('functions.php');
include('simple_html_dom.php');

$save_to_db = true;
$use_local = false;
$readDetails = false;
$localDir = dirname(__FILE__) . '/mirrored_pages/';

$database = array(
	'host' => 'mysql.ryan953.com',
	'login' => 'app_access',
	'password' => 'j85x7Tug',
	'database' => 'com_ryan953_beerprice',
);


if ($save_to_db) {
	getDB($database);
}


pr('Creating a DOM object from search_url');
$search_dom = getDomPage(
	array(
		'url'=>"http://www.thebeerstore.ca/Beers/brandsearch_results.asp?brand=&beertype=-+All+-&category=-+All+-&producer=-+All+-&Submit.x=42&Submit.y=18&Submit=Submit",
		'file'=>"{$localDir}search.html",
		'local'=> $use_local
	)
);

pr('Reading all Beer Names and Details URLs from DOM');
$beers = readBeerNames($search_dom); //read all beer name's from the search_url

pr($beers);
$complete = 0;

foreach ($beers as $name=>$val) {
	pr("Loading {$name}");

	if (!empty($val['details_url']) && $readDetails) {
		pr("Creating DOM object: details");
		$details = getDomPage(
			array(
				'url'=>$val['details_url'],
				'file'=>"{$localDir}details_{$name}.html",
				'local'=> $use_local
			)
		);

		pr("Reading Details");
		$val_details = readBeerDetails($details);
		$details = null;
		$val = array_merge($val, $val_details);
		$val_details = null;
	}

	if (!empty($val['prices_url'])) {
		pr("Creating DOM object: prices");
		$prices = getDomPage(
			array(
				'url'=>$val['prices_url'],
				'file'=>"{$localDir}prices_{$name}.html",
				'local'=> $use_local
			)
		);

		pr("Reading Prices");
		$val_prices = readBeerPrices($prices);
		$prices = null;

		$val = array_merge($val, $val_prices);
		$val_prices = null;
	}

	foreach ($val['prices'] as $i=>$price) {
		$val['prices'][$i] = splitPriceData($price);
	}

	$time = time();

	if (empty($name) || !$save_to_db) {
		continue;
	}

	$beer_id = getId('brands', 'name', $name);

	if (empty($beer_id)) {
		$brewer_id = getId('brewers', 'name', $val['brewer']);
		if (empty($brewer_id)) {
			mysql_query("INSERT INTO brewers (name) VALUES ('{$val['brewer']}')");
			$brewer_id = mysql_insert_id();
		}

		$type_id = getId('types', 'name', $val['type']);
		if (empty($type_id)) {
			mysql_query("INSERT INTO types (name) VALUES ('{$val['type']}')");
			$type_id = mysql_insert_id();
		}

		mysql_query("INSERT INTO brands (name, brewer_id, type_id, percent, year) VALUES ('{$val['name']}', {$brewer_id}, {$type_id}, {$val['percent']}, 0);");
		$beer_id = mysql_insert_id();
	}

	foreach((array)$val['prices'] as $price) {
		//insert price with datestamp
		$package_id = getId('packages', 'name', $price['packageName']);
		if (empty($package_id)) {
			$container_id = getId('containers', 'name', $price['containerName']);
			if (empty($container_id)) {
				mysql_query("INSERT INTO containers (name, volume_amount, volume_unit) VALUES ('{$price['containerName']}', {$price['containerVolumeAmount']}, '{$price['containerVolumeUnit']}')");
				$container_id = mysql_insert_id();
			}

			mysql_query("INSERT INTO packages (name, quantity, container_id) VALUES ('{$price['packageName']}', {$price['containerQuantity']}, {$container_id})");
			$package_id = mysql_insert_id();
		}

		mysql_query("INSERT INTO prices (amount, brand_id, package_id, location_id, timestamp) VALUES ({$price['price']}, {$beer_id}, {$package_id}, 1, UNIX_TIMESTAMP())");
	}

	$count = count($val['prices']);
	$complete++;
	pr("Beer_id {$beer_id} has {$count} prices");
	pr("Completed {$complete}, " . (count($beers) - $complete) . " to go");
}
pr('Done');
?>

File: data-import/getBeer.py
from debug import *

import argparse
import MySQLdb

from BeerStore import Parser as BeerStoreParser

settings = {
    'debugging':True,
    'logging':True,
    'errors':True,
    'cacheDir':'./cache/',
    'searchUrl':"http://www.thebeerstore.ca/beers/search/advanced?brand=All&type=All&sub_type=All&country=All&brewer=All&title_1=",
    'database': {
        'host': 'mysql.ryan953.com',
        'user': 'app_access',
        'passwd': 'j85x7Tug',
        'db': 'com_ryan953_beerprice',
    }
}

class DataSaver(object):
    def __init__(self, dbConf):
        db = MySQLdb.connect(**dbConf)
        self.cursor = db.cursor()
        self.cache = {}

    def save(self, data):
        for dataType in data:
            methodName = 'save'+dataType.capitalize()
            try:
                method = getattr(self, methodName)
            except:
                debug('Count not find: %s' % methodName)
                continue
            method(data[dataType])

    def _defaultSaver(self, table, values):
        self._saveMany(table=table, values=values)
        self._cacheStatics(table=table)

    def _saveMany(self, table, fields=['name'], values=[]):
        f = ", ".join(fields)
        v = ", ".join(["%s"] * len(fields))
        insert = '''INSERT IGNORE INTO %s (%s) VALUES (%s)''' % (table, f, v)
        rows = self.cursor.executemany(insert,  values)
        debug("Inserted %s rows into %s" % (rows, table))

    def _cacheStatics(self, table, unique='id', search='name'):
        if search is None:
            search = table

        select = '''SELECT %s, %s FROM %s''' % (unique, search, table)
        self.cursor.execute(select)
        rows = self.cursor.fetchall()
        self.cache[table] = {}
        for row in rows:
            key = row[1].upper()
            self.cache[table][key] = row[0]
        debug('''Cached: %s''' % (table,))

class BeerSaver(DataSaver):
    def save(self, data):
        brands = data['brands']
        packages = data['packages']
        prices = data['prices']
        del data['brands']
        del data['packages']
        del data['prices']

        DataSaver.save(self, data)
        self.saveBrands(brands)
        self.savePackages(packages)
        self.savePrices(prices)

    def saveCategories(self, values):
        self._defaultSaver('categories', values)

    def saveTypes(self, values):
        self._defaultSaver('types', values)

    def saveStyles(self, values):
        self._defaultSaver('styles', values)

    def saveCountries(self, values):
        self._defaultSaver('countries', values)

    def saveBrewers(self, values):
        self._defaultSaver('brewers', values)

    def saveContainers(self, containers):
        table = 'containers'
        fields = ['name','volume_unit','volume_amount']
        values = [(c['name'], c['volume_unit'], c['volume_amount']) for c in containers]
        self._saveMany(table=table, fields=fields, values=values)
        self._cacheStatics(table=table)

    def saveBrands(self, brands):
        if 'brewers' not in self.cache or 'types' not in self.cache:
            debug('Need Brewers to save Brands')
            return

        for brand in brands:
            typeName = brand['type'].upper()
            brewer = brand['brewer'].upper()
            brand['typeid'] = self.cache['types'][typeName]
            brand['brewerid'] = self.cache['brewers'][brewer]

        table = 'brands'
        fields = ['name', 'percent', 'year', 'type_id', 'brewer_id']
        values = [(b['name'], b['alcohol'], 0, b['typeid'], b['brewerid']) for b in brands]
        self._saveMany(table=table, fields=fields, values=values)
        self._cacheStatics(table=table)

    def savePackages(self, packages):
        if 'containers' not in self.cache:
            debug('Need Contaners to save Packages')
            return

        for package in packages:
            containerName = package['container']['name'].upper()
            package['containerid'] = self.cache['containers'][containerName]

        table = 'packages'
        fields = ['name', 'quantity', 'container_id']
        values = [(p['name'], p['quantity'], p['containerid']) for p in packages]
        self._saveMany(table=table, fields=fields, values=values)
        self._cacheStatics(table=table)

    def savePrices(self, prices):
        if 'brands' not in self.cache or 'packages' not in self.cache:
            debug('Need Brands and Pacakges to save Prices')
            return

        values = []
        for brand in prices:
            for price in brand:
                brandName = price['brand'].upper()
                packageName = price['package']['name'].upper()
                price['brandid'] = self.cache['brands'][brandName]
                price['packageid'] = self.cache['packages'][packageName]
            values.append((price['price'], price['brandid'], price['packageid'], 0))

        table = 'prices'
        fields = ['amount', 'brand_id', 'package_id', 'location_id']
        self._saveMany(table=table, fields=fields, values=values)

class DataImporter(object):
    def scrape(self, args):
        debug(args)
        if args.parse:
            parser = BeerStoreParser(update=args.update, 
                searchUrl=settings['searchUrl'], 
                cacheDir=settings['cacheDir'])
            parsed = parser.parse()
            debug('Data parsed')

            if args.save:
                saver = BeerSaver(settings['database'])
                saver.save(parsed)
        else:
            debug('Nothing to do')

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Populate the BierFrau database with prices from TheBeerStore.ca')
    parser.add_argument('--update', action='store_true', help='update file cache')
    parser.add_argument('--parse', action='store_true', default=True, help='parse cached files')
    parser.add_argument('--save', action='store_true', help='save parsed data')

    importer = DataImporter()
    importer.scrape(parser.parse_args())
